ast.cpp

```diff
@@ -44,47 +44,97 @@ object::object(){
+
 NodeBinOp::NodeBinOp(NodeExpr* l, NodeExpr* r, string p) {
     lhs = l;
     rhs = r;
     punct = p;
     lhs->parent = this;
     rhs->parent = this;
 }
-NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {}
 
-NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {}
+NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {
+    if(lhs->is_NodeId()) {
+        lhs->c_type = rhs->c_type;
+        var_map[((NodeId*)(lhs))->id]->c_type = lhs->c_type;
+    }
+    c_type = rhs->c_type;
+}
 
-NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {}
+NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {
+    c_type = new CIntType();
+}
 
-NodeAdd::NodeAdd(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "+") {}
+NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {
+    if(rhs->c_type->isIntType() && lhs->c_type->isPtrType()) {
+        c_type = new CPtrType(new CIntType());
+    }
+    else{
+        c_type = new CIntType();
+    }
+}
 
-NodeMul::NodeMul(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "*"){}
+NodeAdd::NodeAdd(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "+") {
+    if(rhs->c_type->isIntType() && lhs->c_type->isPtrType()) {
+        c_type = new CPtrType(new CIntType());
+    }
+    else{
+        c_type = new CIntType();
+    }
+}
 
-NodeEE::NodeEE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "=="){}
+NodeMul::NodeMul(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "*"){
+    c_type = new CIntType();
+}
 
-NodeGTE::NodeGTE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, ">="){}
+NodeEE::NodeEE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "=="){
+    c_type = new CIntType();
+}
 
-NodeGT::NodeGT(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, ">"){}
+NodeGTE::NodeGTE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, ">="){
+    c_type = new CIntType();
+}
 
-NodeLTE::NodeLTE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "<="){}
+NodeGT::NodeGT(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, ">"){
+    c_type = new CIntType();
+}
 
-NodeLT::NodeLT(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "<"){}
+NodeLTE::NodeLTE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "<="){
+    c_type = new CIntType();
+}
 
-NodeNE::NodeNE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "!="){}
+NodeLT::NodeLT(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "<"){
+    c_type = new CIntType();
+}
+
+NodeNE::NodeNE(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "!="){
+    //c_type = new CIntType();
+}
 
 NodeExprStmt::NodeExprStmt(NodeExpr* e){
     _expr = e;
     _expr->parent = this;
 };
 
+CIntType::CIntType(){
+    size = 8;
+};
+
+CPtrType::CPtrType(CType* r){
+    referenced_type = r;
+    size = 8;
+};
+
 NodeNum::NodeNum(int n){
     num_literal = n;
+    c_type = new CIntType();
 }
 
 NodeAddressOf::NodeAddressOf(NodeExpr* e){
     _expr = e;
     _expr->parent = this;
+    c_type = new CPtrType(_expr->c_type);
 }
 
 NodeDereference::NodeDereference(NodeExpr* e) {
```

Comment:
NodeDereference did not set its c_type. The expression's type in a NodeDereference should have at least 1 level of pointer indirection, but can have unlimited pointer indirections. The result is one less pointer indirection than what it had. After copying `_expr = e`, do
```
assert(_expr->c_type->isPtrType() && "* deref operator requires a pointer");
```
Then since we know it is a pointer type (otherwise the assert would crash the program), we can safely cast to `_expr`'s c_type field to a CPtrType and get its reference type, which is one less level of indirection like we need.
```
c_type = ((CPtrType*)(_expr->c_type))->referenced_type;
```
A good test for this could be 
```
assert 5 '{ a=3; b=5; c=&a; d=&c; return *((*d) + 1);}'
```
d is a pointer to a pointer to an int, so dereferencing it should give you a pointer to an int. Then doing +1 on that should result in pointer arithmetic, giving you the location of b. Then dereferencing that will give you the value of b.